### PR TITLE
KEYCLOAK-12194 - Max SSO idle timeout greater than 24 days cause integer overflow on infinispan cache entry expiration

### PR DESCRIPTION
**Pull request by alexandredantas**
As described on Jira ticket, when a remote cache is set up, expiration can overflow when max idle times are greater than 24 days.

It happens because times on keycloak are stored as seconds but when writing to remote cache, it's multiplied by 1000, causing an overflow.

### CODE REVIEW

**Comment by alexandredantas** [comment]
I didn't find an way to test remote cache. If there are any guidelines for this, please let me know

File: model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/InfinispanUserSessionProviderFactory.java
Comment (mposolda) [inline model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/InfinispanUserSessionProviderFactory.java:251]: I don't think that adding "Time.currentTimeMillis()" is correct here?

This is idle timeout. Hence if "Time.currentTimeMillis" is added, it means the expiration won't be around 10 hours (as default SSO session max lifespan), but around 40+ years (Time since 1 Jan 1970). Use something like `(1000L * realm.getSsoSessionMaxLifespan()` should do the trick with retype to long though?

**Review by mposolda — CHANGES_REQUESTED** [review]
@alexandredantas Thanks for the PR and sorry for late review. Added inline comment.

File: model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/InfinispanUserSessionProviderFactory.java
Comment (alexandredantas) [reply model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/InfinispanUserSessionProviderFactory.java:251]: Hi Marek! No problems :)

Actually, when a remote Infinispan cache is set up the model expects an epoch, not a lifespan, this is why adding Time.currentTimeMillis solves the issue. I'm using `org.keycloak.cluster.infinispan.KeycloakHotRodMarshallerFactory` to marshall data from keycloak to hot rod server

Comment (mposolda) [reply model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/InfinispanUserSessionProviderFactory.java:251]: Hi Alexandre,

I am not sure how you figured this statement? When looking at the javadoc of method org.infinispan.commons.api.BasicCache and method `put(K key, V value, long lifespan, TimeUnit lifespanUnit, long maxIdleTime, TimeUnit maxIdleTimeUnit)` , it mentions this for the maxIdleTime field: the maximum amount of time this key is allowed to be idle for before it is considered as expired .

This is the method, which Keycloak internally calls. So I don't understand why this is supposed to be the epoch? Any more info about this?

Comment (alexandredantas) [reply model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/InfinispanUserSessionProviderFactory.java:251]: I've setup my environment the like described below:

1 - An external Infinispan Hot Rod Server with JDBC store configuration
```
Configuration persistentCacheConfiguration = new ConfigurationBuilder()
                    .memory().size(100).evictionStrategy(EvictionStrategy.REMOVE)
                    .persistence().addStore(JdbcStringBasedStoreConfigurationBuilder.class)
                    .table()
                    .idColumnName("id")
                    .idColumnType("varchar(1000)")
                    .dataColumnName("data")
                    .dataColumnType("bytea")
                    .timestampColumnName("expire")
                    .timestampColumnType("bigint")
                    .dropOnExit(false)
                    .createOnStart(true)
                    .tableNamePrefix("ispn")
                    .connectionPool()
                    .connectionUrl(cfg.get("DATABASE_JDBC_URL"))
                    .driverClass(cfg.get("DATABASE_JDBC_DRIVER"))
                    .password(cfg.get("DATABASE_PASSWORD"))
                    .username(cfg.get("DATABASE_USERNAME"))
                    .build();

            DefaultCacheManager defaultCacheManager = new DefaultCacheManager(
                    new GlobalConfigurationBuilder()
                            .defaultCacheName("sessions")
                            .build(), persistentCacheConfiguration);

            defaultCacheManager.createCache("___script_cache", new ConfigurationBuilder().build());
            defaultCacheManager.createCache("clientSessions", persistentCacheConfiguration);
            defaultCacheManager.createCache("offlineClientSessions", persistentCacheConfiguration);
            defaultCacheManager.createCache("offlineSessions", persistentCacheConfiguration);
            defaultCacheManager.createCache("actionTokens", persistentCacheConfiguration);
            defaultCacheManager.createCache("loginFailures", persistentCacheConfiguration);

            new HotRodServer().start(new HotRodServerConfigurationBuilder().host("127.0.0.1").port(11322).build(),
                    defaultCacheManager);
```

2 - Remote socket configuration
```
/socket-binding-group=standard-sockets/remote-destination-outbound-socket-binding=remote-cache:add(host="127.0.0.1", port="11322")
```

3 - Cache configuration

```
/subsystem=infinispan/cache-container=keycloak/local-cache=sessions/store=remote:add(cache="sessions", remote-servers=[remote-cache], preload=false, purge=false, passivation=false, shared=false, properties={rawValues=true, marshaller=org.keycloak.cluster.infinispan.KeycloakHotRodMarshallerFactory, protocolVersion=${env.HOTROD_VERSION:2.8}}){allow-resource-service-restart=true}
/subsystem=infinispan/cache-container=keycloak/local-cache=clientSessions/store=remote:add(cache="clientSessions", remote-servers=[remote-cache], preload=false, purge=false, passivation=false, shared=false, properties={rawValues=true, marshaller=org.keycloak.cluster.infinispan.KeycloakHotRodMarshallerFactory, protocolVersion=${env.HOTROD_VERSION:2.8}}){allow-resource-service-restart=true}
/subsystem=infinispan/cache-container=keycloak/local-cache=offlineClientSessions/store=remote:add(cache="offlineClientSessions", remote-servers=[remote-cache], preload=false, purge=false, passivation=false, shared=false, properties={rawValues=true, marshaller=org.keycloak.cluster.infinispan.KeycloakHotRodMarshallerFactory, protocolVersion=${env.HOTROD_VERSION:2.8}}){allow-resource-service-restart=true}
/subsystem=infinispan/cache-container=keycloak/local-cache=offlineSessions/store=remote:add(cache="offlineSessions", remote-servers=[remote-cache], preload=false, purge=false, passivation=false, shared=false, properties={rawValues=true, marshaller=org.keycloak.cluster.infinispan.KeycloakHotRodMarshallerFactory, protocolVersion=${env.HOTROD_VERSION:2.8}}){allow-resource-service-restart=true}
/subsystem=infinispan/cache-container=keycloak/local-cache=authenticationSessions/store=remote:add(cache="authenticationSessions", remote-servers=[remote-cache], preload=false, purge=false, passivation=false, shared=false, properties={rawValues=true, marshaller=org.keycloak.cluster.infinispan.KeycloakHotRodMarshallerFactory, protocolVersion=${env.HOTROD_VERSION:2.8}}){allow-resource-service-restart=true}
```

After performing a login on Keycloak, I retrieved some data from database table ispn_clientSessions. There the timestamp field is not a max idle but an epoch

**Comment by mposolda** [comment]
@alexandredantas Ah, Thanks. So now I understand how you figured it.

However the timestamp on the ispn_clientSessions table has nothing to do with the parameters of "remoteCache.put" method. You can try for example something like this:
```
        remoteCache1.put("key1", "val1", 1200, TimeUnit.SECONDS, 60, TimeUnit.SECONDS);
        remoteCache1.put("key2", "val2", 1200, TimeUnit.SECONDS, 30, TimeUnit.SECONDS);

        try {
            Thread.sleep(45000);
        } catch (InterruptedException ie) {
            throw new RuntimeException(ie);
        }

        // Key1 will be still active. Key2 will be expired.
        System.out.println("key1: " + remoteCache1.get("key1") + ", key2: " + remoteCache1.get("key2"));
```

See infinispan docs/quickstarts for some more example how to make remoteCache up and running if you want to try something like this.

**Comment by alexandredantas** [comment]
@mposolda Right, timestamp column gets data from `maxIdleLoader` passed on method
```
public void addRemoteCache(String cacheName, RemoteCache remoteCache, MaxIdleTimeLoader maxIdleLoader)
```
on `RemoteCacheInvoker` class.

Once this time is exceeded, a cleanup job on Keycloak removes sessions. What I noticed was timestamp column is an epoch, not a max idle.

I found these issue when suddenly our persisted sessions simply disappeared from the database and debugging I was able to find only this point where timestamp column is populated.

**Comment by alexandredantas** [comment]
@mposolda is there anything I can help about this issue?

**Comment by mposolda** [comment]
> @mposolda Right, timestamp column gets data from `maxIdleLoader` passed on method
> 
> ```
> public void addRemoteCache(String cacheName, RemoteCache remoteCache, MaxIdleTimeLoader maxIdleLoader)
> ```
> 
> on `RemoteCacheInvoker` class.
> 
> Once this time is exceeded, a cleanup job on Keycloak removes sessions. What I noticed was timestamp column is an epoch, not a max idle.
> 
> I found this issue when suddenly our persisted sessions simply disappeared from the database and debugging I was able to find only this point where timestamp column is populated.
> 
> I see your point about Infinispan and I agree that Infinispan has eviction policy by time, as you pointed out but this issue is not exactly related to eviction. It's more related to cleanup job based on timestamp field.

@alexandredantas Good afternoon and sorry for late response. In the middle of PTO season here :-)

I am pretty sure that currently this PR is not correct and will require some rewrite as at least some timestamps here don't refer to an "epoch", but to an "interval" . This can't be accepted as we need to make sure that entries are regularly expired from the remoteCache eventually after 10 days (time specified by (realm.getSsoSessionMaxLifespan() * 1000) milliseconds), not after 50 years :-)

For more details: In the PR you changed this among other things:
```
 boolean sessionsRemoteCache = checkRemoteCache(session, sessionsCache, (RealmModel realm) -> {
            // We won't write to the remoteCache during token refresh, so the timeout needs to be longer.
            return realm.getSsoSessionMaxLifespan() * 1000;
            return Time.currentTimeMillis() + (realm.getSsoSessionMaxLifespan() * 1000L);
```

The method "checkRemoteCache" calls this line
```
remoteCacheInvoker.addRemoteCache(ispnCache.getName(), remoteCache, maxIdleLoader);
```

This remoteCache invoker calls this line:
```
long loadedMaxIdleTimeMs = context.maxIdleTimeLoader.getMaxIdleTimeMs(realm);
```
And this calls method "runOnRemoteCache", which calls for example this
```
remoteCache.put(key, sessionWrapper.forTransport(), task.getLifespanMs(), TimeUnit.MILLISECONDS, maxIdleMs, TimeUnit.MILLISECONDS);
```

The "maxIdleMs" argument passed on the remoteCache.put is definitely supposed to be "interval" and not an "epoch" . However you changed it to the epoch, which means that entry will be eventually expired from the cache after around 50 years instead of 10 days.

Again, I suggest to try this:
```
        remoteCache1.put("key1", "val1", 1200, TimeUnit.SECONDS, 60, TimeUnit.SECONDS);
        remoteCache1.put("key2", "val2", 1200, TimeUnit.SECONDS, 30, TimeUnit.SECONDS);

        try {
            Thread.sleep(45000);
        } catch (InterruptedException ie) {
            throw new RuntimeException(ie);
        }

        // Key1 will be still active. Key2 will be expired.
        System.out.println("key1: " + remoteCache1.get("key1") + ", key2: " + remoteCache1.get("key2"));
```

Please let me know if I was inaccurate in some statement above or please update this PR.

**Comment by alexandredantas** [comment]
Hi @mposolda no problems :)

Actually, I tried already the approach you suggested and it didn't work. To troubleshoot the issue, I've created a column on Infinispan created table to record when it was inserted and using this column I'm able to compare if session expiration is ok.

Currenty, our realm has an `SSO session max` set to 365 days. Below, there is a record on database with the proposed fix:

```
timestamp		created_at
1629478657962	2020-08-20 16:27:37.961275
```

When comparing the timestamp + 365 days with created_at + 365 days, it matches.

Maybe there is an issue at Infinispan side and then I agree with you that this fix is not related to Keycloak

**Comment by mposolda** [comment]
@alexandredantas Thanks! I hope I will be able to merge this once our travis passed. Just re-run the tests...

**Comment by alexandredantas** [comment]
@mposolda tks! how do I re-run tests?

**Comment by mposolda** [comment]
@alexandredantas I've tried to re-run the tests few times, however they are still failing in travis. Namely the SessionExpirationCrossDCTest, SessionsPreloadCrossDCTest and LastSessionRefreshCrossDCTest are failing. Do you have a chance to try to run them localy with the Keycloak on Wildfly/Undertow server and check if your changes caused any regression?  The instructions are here https://github.com/keycloak/keycloak/blob/master/testsuite/integration-arquillian/HOW-TO-RUN.md#cross-dc-tests .

**Comment by alexandredantas** [comment]
@mposolda I'd performed a rebase on my branch and ran the tests but I was only able to run them after adding `--add-modules=java.se` on arquillian.xml `javaVmArguments`. After running the suites you've mentioned, I've got the following errors:

SessionExpirationCrossDCTest and LastSessionRefreshCrossDCTest:
```
java.lang.NoClassDefFoundError: Could not initialize class org.jboss.marshalling.river.RiverMarshaller
```

SessionExpirationCrossDCTest and LastSessionRefreshCrossDCTest:
```
Caused by: java.lang.IllegalStateException: No standard field found for reverse order comparator!
	at org.jboss.marshalling.river.Protocol.<clinit>(Protocol.java:287)
	... 136 more
```

SessionsPreloadCrossDCTest:
```
"ISPN000299: Unable to acquire lock after 0 milliseconds for key"
```

I'm not sure if my change was capable of introducing this kind of classloading issues. I'd tried java 11 and 14. Is there anything else I can try or should change on modules to make it work?

**Comment by mposolda** [comment]
@alexandredantas I suggest that you possibly squash the 2 commits in this PR to the single commit and then you rebase this PR to the latest master. Maybe the rebase will help to fix the tests...

**Comment by alexandredantas** [comment]
@mposolda when I run cross dc tests on master, I'm getting the same errors as those on my branch. I'll re-check docs about running tests to check.

By the way, after rebasing and squashing, other tests failed on Travis, not those cross dc.

**Comment by mposolda** [comment]
@alexandredantas I see. I re-run the tests in travis, but it looks that rebase did not help. I am trying to run in our lab environment too to see if the failed tests are regressions of your changes.

**Comment by mposolda** [comment]
@alexandredantas I've run the cross-dc tests in our CI environment. Unfortunately there are failures after your changes :( The failures were:
```
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   SessionExpirationCrossDCTest.testClearDetachedOfflineClientSessions:580->createInitialSessions:163->lambda$createInitialSessions$0:173 expected:<7> but was:<20>
[ERROR]   SessionExpirationCrossDCTest.testPeriodicExpirationOfflineSessions:304->createInitialSessions:163->lambda$createInitialSessions$0:173 expected:<12> but was:<20>
[ERROR]   SessionExpirationCrossDCTest.testRealmRemoveOfflineSessions:222->createInitialSessions:163->lambda$createInitialSessions$0:173 expected:<10> but was:<20>
[ERROR]   SessionExpirationCrossDCTest.testUserRemoveOfflineSessions:374->createInitialSessions:163->lambda$createInitialSessions$0:173 expected:<10> but was:<20>
[ERROR]   SessionsPreloadCrossDCTest.offlineSessionsPreloadTest:103 expected:<2> but was:<10>
[ERROR] Errors: 
[ERROR]   LastSessionRefreshCrossDCTest.testOfflineSessionsLastSessionRefreshUpdate:279->assertStatistics:361->lambda$assertStatistics$0:368 » NotFound
[INFO] 
[ERROR] Tests run: 37, Failures: 5, Errors: 1, Skipped: 0
```

In other words, the PR can't be accept as is due the fact that it introduces regressions and test failures. Test failures would need to be fixed before the PR can be accepted.

**Comment by alexandredantas** [comment]
@mposolda how can I run cross dc tests on my machine? I tried the way described on (how to)[https://github.com/keycloak/keycloak/blob/master/testsuite/integration-arquillian/HOW-TO-RUN.md#cross-dc-tests] and got different errors. With java 8 I've got tests stuck during the run and with java 11 I've got issues with `RiverMarshaller` or timeout on infinispan.
Sounds very weird this simple change breaking tests this way. Is there anything I can do to proceed with this PR?

**Comment by mposolda** [comment]
@alexandredantas Hopefully these 3 commands are sufficient. I suggest OpenJDK 8:
```
mvn clean install -DskipTests=true -Pdistribution
mvn -Pcache-server-infinispan,auth-servers-crossdc-jboss,auth-server-wildfly -f testsuite/integration-arquillian -DskipTests clean install
mvn -Pcache-server-infinispan,auth-servers-crossdc-jboss,auth-server-wildfly -Dtest=org.keycloak.testsuite.crossdc.**.*Test -pl testsuite/integration-arquillian/tests/base clean install
```
If you see the different errors, it may be caused by the fact that some tests affected the other tests. It is possible to maybe try to run the single test from those tests, which failed. So for example for `-Dtest` you will use something like `-Dtest=LastSessionRefreshCrossDCTest` and see if it works. You can also try to run this without your PR and then after your PR to doublecheck if you see the regression on your laptop with/without your changes.

> Is there anything I can do to fix and proceed with this PR?

Fix the tests :/ And as you can imagine, sometimes fixing the regression in the test can be much bigger effort than fixing the issue itself. It is not very often case, but it seems that your PR is one of those rare cases where it can be the case :(

**Comment by mposolda** [comment]
@alexandredantas I am sorry for closing this for now as it is pending for a long time without more feedback... I appreciate the work you put into this, but unfortunately, we can't merge the PR, which introduces regression to the testsuite and failing test. Feel free to re-open once you figure how to fix the testsuite and the cause why the test are failing.

**Pull request closed without merging**
